File: Chapter10/C#/RunningSamples/EncoderMedia/MediaPlayer/ThumbnailDownloader.cs
// <copyright file="ThumbnailDownloader.cs" company="Microsoft">
//     Copyright © Microsoft Corporation. All rights reserved.
// </copyright>
// <summary>Implements the ThumbnailDownloader class</summary>
// <author>Microsoft Expression Encoder Team</author>
using System;
using System.Collections.Generic;
using System.Windows.Threading;

namespace ExpressionMediaPlayer
{
    public sealed class ThumbnailDownloader
    {
        /// <summary>
        /// Dispatch timer for downloading thumbnails.
        /// </summary>
        private static DispatcherTimer sm_timerThumbnails;
        /// <summary>
        /// queue of thumbnails to download
        /// </summary>
        private static Queue<ThumbnailImage> sm_downloadQueue = new Queue<ThumbnailImage>();
        /// <summary>
        /// create ThumbnailDownloader
        /// </summary>
        private ThumbnailDownloader()
        {
        }
        /// <summary>
        /// Event handler for a Timer tick.
        /// </summary>
        /// <param name="sender">Source of the event.</param>
        /// <param name="e">Event args.</param>
        private static void OnTimerTickThumbnailDownload(object sender, EventArgs e)
        {
            lock (sm_downloadQueue)
            {
                if (sm_downloadQueue.Count > 0)
                {
                    ThumbnailImage thumbnail = sm_downloadQueue.Dequeue();
                    thumbnail.LoadImage();
                }
            }
        }        
        /// <summary>
        /// Creates the timer for downloading thumbnails.
        /// </summary>
        public static void EnableThumbnailDownload()
        {
            if (sm_timerThumbnails == null)
            {
                sm_timerThumbnails = new DispatcherTimer();
                sm_timerThumbnails.Interval = new TimeSpan(0/*days*/, 0/*hours*/, 0/*minutes*/, 0/*seconds*/, 60/*milliseconds*/);
                sm_timerThumbnails.Tick += new EventHandler(OnTimerTickThumbnailDownload);
            }

            sm_timerThumbnails.Start();
        }
        /// <summary>
        /// Turns off the timer for downloading thumbnails.
        /// </summary>
        public static void DisableThumbnailDownload()
        {
            if (sm_timerThumbnails != null)
            {
                sm_timerThumbnails.Stop();
            }
        }
        /// <summary>
        /// Place a thumbnail download request in the queue.
        /// </summary>
        public static void Enqueue(ThumbnailImage thumbnail)
        {
            lock (sm_downloadQueue)
            {
                sm_downloadQueue.Enqueue(thumbnail);
            }
        }
    }
}

File: Chapter10/VB/VBProject_10/AdvancedMedia/1-AdvanceMedia/AdvMediaPlayer.Web/EncoderMedia/MediaPlayer/ThumbnailDownloader.vb
'  <copyright file="ThumbnailDownloader.cs" company="Microsoft">
'      Copyright © Microsoft Corporation. All rights reserved.
'  </copyright>
'  <summary>Implements the ThumbnailDownloader class</summary>
'  <author>Microsoft Expression Encoder Team</author>
Imports System
Imports System.Collections.Generic
Imports System.Windows.Threading

' Namespace ExpressionMediaPlayer

    Public sealed Class ThumbnailDownloader
        ''' <summary>
        ''' Dispatch timer for downloading thumbnails.
        ''' </summary>
        Private ReadOnly Property DispatcherTimer() As
        ''' <summary>
        ''' Event handler for a Timer tick.
        ''' </summary>
        ''' <param name="sender">Source of the event.</param>
        ''' <param name="e">Event args.</param>
        End Property


            lock (sm_downloadQueue)
            {

                If (sm_downloadQueue.Count > 0) Then
                    ThumbnailImage thumbnail = sm_downloadQueue.Dequeue()
                    thumbnail.LoadImage()
                End If

            }
        End Sub '   OnTimerTickThumbnailDownload
        ''' <summary>
        ''' Creates the timer for downloading thumbnails.
        ''' </summary>
        End Property



            If (sm_timerThumbnails Is Nothing) Then
                sm_timerThumbnails = New DispatcherTimer()
                sm_timerThumbnails.Interval = New TimeSpan(0/*days*/, 0/*hours*/, 0/*minutes*/, 0/*seconds*/, 60/*milliseconds*/)
                AddHandler sm_timerThumbnails.Tick, AddressOf EventHandler(OnTimerTickThumbnailDownload)
            End If


            sm_timerThumbnails.Start()
        End Sub '   EnableThumbnailDownload
        ''' <summary>
        ''' Turns off the timer for downloading thumbnails.
        ''' </summary>
        End Property



            If (sm_timerThumbnails  IsNot Nothing) Then
                sm_timerThumbnails.Stop()
            End If
        End Sub '   DisableThumbnailDownload
        ''' <summary>
        ''' Place a thumbnail download request in the queue.
        ''' </summary>
        End Property


            lock (sm_downloadQueue)
            {
                sm_downloadQueue.Enqueue(thumbnail)
            }
        End Sub '   Enqueue
    End Class   '   ThumbnailDownloader
' End Namespace   '   ExpressionMediaPlayer
' ..\Project_10\AdvancedMedia\1-AdvanceMedia\AdvMediaPlayer.Web\EncoderMedia\MediaPlayer\ThumbnailDownloader.cs
